codegen/post_gen_hooks.go

```diff
@@ -23,102 +23,204 @@ package codegen
+							instance.InstanceName,
+						)
+						return
+					}
+
+					files[filepath.Join(genDir, "types.go")] = types
+					files[filepath.Join(genDir, "mock_client_with_fixture.go")] = augMock
+				}
+
+				fmt.Printf(
+					genFormattor,
+					"mock",
+					instance.ClassName,
 					instance.InstanceName,
+					path.Join(path.Base(buildDir), instance.Directory, "mock-client"),
+					i+1, mockCount,
 				)
+			}(instance, i)
+		}
+		wg.Wait()
+
+		select {
+		case err := <-ec:
+			return err
+		default:
+		}
+
+		for p, data := range files {
+			if err := writeAndFormat(p, data); err != nil {
+				return err
 			}
-			fmt.Printf(
-				"Generating %12s %12s %-30s in %-50s %d/%d\n",
-				"mock",
-				instance.ClassName,
-				instance.InstanceName,
-				path.Join(path.Base(h.CodeGenTargetPath()), instance.Directory, "mock-client"),
-				i+1, mockCount,
-			)
 		}
 		return nil
 	}, nil
 }
 
-// ServiceMockGenHook returns a PostGenHook to generate server mocks
+// ServiceMockGenHook returns a PostGenHook to generate service mocks
 func ServiceMockGenHook(h *PackageHelper, t *Template) PostGenHook {
 	return func(instances map[string][]*ModuleInstance) error {
 		fmt.Println("Generating service mocks:")
 		mockCount := len(instances["service"])
+		files := make(map[string][]byte)
+		ec := make(chan error, mockCount)
+		var wg sync.WaitGroup
+		wg.Add(mockCount)
 		for i, instance := range instances["service"] {
-			mockInit, err := generateMockInitializer(instance, h, t)
-			if err != nil {
-				return errors.Wrapf(
-					err,
-					"Error generating service mock_init.go for %s",
-					instance.InstanceName,
-				)
-			}
-			mockService, err := t.ExecTemplate("service_mock.tmpl", instance, h)
-			if err != nil {
-				return errors.Wrapf(
-					err,
-					"Error generating service mock_service.go for %s",
-					instance.InstanceName,
-				)
-			}
+			go func(instance *ModuleInstance, i int) {
+				defer wg.Done()
+
+				buildDir := h.CodeGenTargetPath()
+				genDir := filepath.Join(buildDir, instance.Directory, "mock-service")
 
-			buildPath := filepath.Join(h.CodeGenTargetPath(), instance.Directory)
-			mockInitPath := filepath.Join(buildPath, "mock-service/mock_init.go")
-			mockServicePath := filepath.Join(buildPath, "mock-service/mock_service.go")
-
-			for _, s := range []struct {
-				path    string
-				content []byte
-			}{
-				{mockInitPath, mockInit},
-				{mockServicePath, mockService},
-			} {
-				if err := writeFile(s.path, s.content); err != nil {
-					return errors.Wrapf(
+				mockInit, err := generateMockInitializer(instance, h, t)
+				if err != nil {
+					ec <- errors.Wrapf(
 						err,
-						"Error writing to file %q",
-						s.path,
+						"Error generating service mock_init.go for %s",
+						instance.InstanceName,
 					)
+					return
 				}
-				if err := FormatGoFile(s.path); err != nil {
-					return err
+				files[filepath.Join(genDir, "mock_init.go")] = mockInit
```

Comment:
should probably use https://golang.org/doc/go1.9#sync-map

Reply:
good catch, replaced map with sync.Map